### PR TITLE
Parachute plugin for simulating flight termination in gazebo

### PR DESCRIPTION
**Pull request by Jaeyoung-Lim**
This PR adds a parachute plugin which is used to simulate the flight termination in gazebo sitl.

The parachute model the plugin comes from https://github.com/AurelienRoy/ardupilot_sitl_gazebo_plugin but integrated into the px4 sitl_gazebo

Since the original code is distributed under Apache, I am not sure if there needs to be a more correct license header in the parachute plugin. The code itself needed to be modified quite a lot to be able to work nicely with the px4 HIL interface.

Below is the demo of the parachute plugin working with the roll angle exceeding the limit.

Video can be seen in the link: [Video](https://youtu.be/17x3Y7K8-j0)

### CODE REVIEW

File: models/parachute_small/model.config [hunk 1]
```diff
@@ -0,0 +1,18 @@
+<?xml version="1.0"?>
+
+<model>
+  <name>Parachute small</name>
+  <version>1.0</version>
+  <sdf version="1.5">model.sdf</sdf>
+
```
Comment (TSC21) [inline models/parachute_small/model.config]: You can add the license as:
```suggestion
  <license>Apache License, Version 2.0</license>
```

File: include/gazebo_parachute_plugin.h [hunk 2]
```diff
@@ -0,0 +1,88 @@
...
+  physics::WorldPtr world_;
+  event::ConnectionPtr _updateConnection;
+
+  bool attached_parachute_ = false;
+  double max_rot_velocity_ = 3500;
+  double ref_motor_rot_vel_;
+  double terminate_rot_vel_ = -1700;
+  int motor_number_;
+
+  std::string trigger_sub_topic_ = "~/video_stream";
```
Comment (TSC21) [inline include/gazebo_parachute_plugin.h]: Why is the trigger the `video_stream` topic?

**Review by TSC21 — CHANGES_REQUESTED** [review]

**Comment by Jaeyoung-Lim** [comment]
@TSC21 Thanks for the review!

**Comment by LorenzMeier** [comment]
Please rebase to re-run testing.

File: include/gazebo_parachute_plugin.h [hunk 3]
```diff
@@ -0,0 +1,88 @@
...
+
+
+  std::string namespace_;
+  physics::ModelPtr model_;
+  physics::WorldPtr world_;
+  event::ConnectionPtr _updateConnection;
+
+  bool attached_parachute_ = false;
+  double max_rot_velocity_ = 3500;
+  double ref_motor_rot_vel_;
```
Comment (TSC21) [inline include/gazebo_parachute_plugin.h]:
```suggestion
  double ref_motor_rot_vel_ = 0;
```

File: include/gazebo_parachute_plugin.h [hunk 4]
```diff
@@ -0,0 +1,88 @@
...
+  void VelocityCallback(CommandMotorSpeedPtr &rot_velocities);
+  void AttachParachute(physics::ModelPtr &parachute_model);
+
+
+  std::string namespace_;
+  physics::ModelPtr model_;
+  physics::WorldPtr world_;
+  event::ConnectionPtr _updateConnection;
+
+  bool attached_parachute_ = false;
```
Comment (TSC21) [inline include/gazebo_parachute_plugin.h]: Can you add some comments on what each one of these variables is doing? Please use `///<`.

File: include/gazebo_parachute_plugin.h [hunk 5]
```diff
@@ -0,0 +1,88 @@
...
+public:
+  ParachutePlugin();
+  virtual ~ParachutePlugin();
+
+protected:
+  virtual void Load(physics::ModelPtr model, sdf::ElementPtr sdf);
+  virtual void OnUpdate(const common::UpdateInfo&);
+
+private:
+  void LoadParachute();
```
Comment (TSC21) [inline include/gazebo_parachute_plugin.h:65]: Some comments here can be used as well:
Ex:
```c++
  /// \brief Trigger the parachute
  /// \param _msg msg received to trigger the parachute
  void TriggerCallback(const boost::shared_ptr<const msgs::Int> &_msg);
```

File: models/if750a/if750a.sdf [hunk 6]
```diff
@@ -494,67 +499,13 @@
...
-            <iMin>0</iMin>
-            <cmdMax>20</cmdMax>
-            <cmdMin>-20</cmdMin>
-          </joint_control_pid>
-        </channel>
-        <channel name='rotor8'>
-          <input_index>7</input_index>
-          <input_offset>0</input_offset>
-          <input_scaling>0.524</input_scaling>
+          <input_scaling>1000</input_scaling>
```
Comment (TSC21) [inline models/if750a/if750a.sdf:505]: Can you explain the rationale behind this?

File: src/gazebo_parachute_plugin.cpp [hunk 7]
```diff
@@ -0,0 +1,131 @@
...
+
+  msgs::Int request;
+  request.set_data(0);
+  
+}
+
+void ParachutePlugin::AttachParachute(physics::ModelPtr &parachute_model){
+
+#if GAZEBO_MAJOR_VERSION >= 9
+  const ignition::math::Pose3d uavPose = model_->WorldPose();
```
Comment (TSC21) [inline src/gazebo_parachute_plugin.cpp]: Can we change the naming from `uavPose` to `vehicle_pose`? Also, the naming format doesn't seem to be uniform (some camel case, then some snake case). Let's keep the variables snake case, unless you have another suggestion.

**Review by TSC21 — CHANGES_REQUESTED** [review]
@Jaeyoung-Lim just some minor stuff. Otherwise good to go.

File: models/if750a/if750a.sdf (hunk 6 above)
Comment (Jaeyoung-Lim) [reply models/if750a/if750a.sdf:505]: @TSC21 I just wanted to use the same scale as other channels. Was this the question?

File: src/gazebo_parachute_plugin.cpp (hunk 7 above)
Comment (Jaeyoung-Lim) [reply src/gazebo_parachute_plugin.cpp]: Done!

**Comment by hamishwillee** [comment]
@Jaeyoung-Lim When merged can you add docs? Maybe something in https://dev.px4.io/master/en/simulation/gazebo.html#usageconfiguration-options ?
Possibly also a link to that information from https://docs.px4.io/master/en/peripherals/parachute.html and

**Comment by Jaeyoung-Lim** [comment]
@hamishwillee Sure! 
@TSC21 Is there anything else missing?

**Comment by Jaeyoung-Lim** [comment]
@TSC21 Is there anything else to fix?

**Review by TSC21 — APPROVED** [review]

**Comment by hamishwillee** [comment]
Hi @Jaeyoung-Lim This merged 4 minutes ago, where are the docs :-)

What we need is something like https://dev.px4.io/master/en/simulation/gazebo.html#simulated-survey-camera

Essentially how do you start the simulation, and everything you need to use it in terms of PX4 setup. I presume this will link both ways to/from:
- https://docs.px4.io/master/en/peripherals/parachute.html
- https://docs.px4.io/master/en/advanced_config/flight_termination.html

Thoughts?

**Comment by SrikarSiddarth** [comment]
Query: I am a student and a beginner in px4, could you please let me know how to trigger the parachute? I know how to launch the if750a in an empty world, but I am unable to know what is the exact condition upon the parachute is triggered and how to command it. Thanks in advance

**Comment by Jaeyoung-Lim** [comment]
@SrikarSiddarth To answer your question, you can look at http://docs.px4.io/master/en/simulation/gazebo.html#simulated-parachute-flight-termination 

Please either post a new issue, or post your question in discuss.px4.io, since your question is not related to this PR

**Comment by SrikarSiddarth** [comment]
Thanks for the quick reply!

**Pull request merged**